### PR TITLE
fix(NcAppNavigation): add focus trap on mobile and improve a11y

### PR DESCRIPTION
**Pull request by ShGKme**
### ☑️ Resolves

* A part of https://github.com/nextcloud/server/issues/36964

### 🖼️ Screenshots

🏚️ Before | 🏡 After
---|---
nav-before | nav-after

### 🚧 Tasks

- [x] Add focus trap, and activate it on **open** navigation on **mobile**, deactivate otherwise
- [x] Improve a11y a bit:
  - [x] Use `<nav>` instead of `role="navigation"` cc @JuliaKirschenheuter to confirm that this is fine
  - [x] Move the toggle button out from the navigation
  - [x] Add `ariaLabel` and `ariaLabelledBy` props to be set on navigation
- [x] Close navigation by <kbd>ESC</kbd> on mobile
- [x] Update tests
  - [x] Test by class instead of internal data property (testing `is closed or open` state by class also is not great, but we should not test by internal implementation as data property is not a part of the component interface). Alternative: `aria-hidden`.
  - [x] Add mobile state test
  - [x] Add a11y tests

### 🏁 Checklist

- [x] ⛑️ Tests are included or are not applicable
- [x] 📘 Component documentation has been extended, updated or is not applicable

### CODE REVIEW

**Comment by szaimen** [comment]
node tests seems to be failing

File: src/components/NcAppNavigation/NcAppNavigation.vue [hunk 1]
```diff
@@ -135,9 +168,27 @@ export default {
...
+			if (this.isMobile && this.open) {
+				this.focusTrap.activate()
+			} else {
+				this.focusTrap.deactivate()
+			}
+		},
+
+		handleEsc() {
+			if (this.isMobile) {
+				this.toggleNavigation(false)
```
Comment (szaimen) [inline src/components/NcAppNavigation/NcAppNavigation.vue:189]: This does not seem to work currently in my testing

**Review by szaimen — COMMENTED** [review]
Tested and the focus trap seems to work :)

However the ESC key doesnt toggle the navigation currently


https://github.com/nextcloud-libraries/nextcloud-vue/assets/42591237/9037fba3-e4ff-4d16-a683-2ffe132fc45f

File: src/components/NcAppNavigation/NcAppNavigation.vue (hunk 1 above)
Comment (szaimen) [reply src/components/NcAppNavigation/NcAppNavigation.vue:189]: the reason was probably that I tested on a screen with higher res than 1024px (so ismobile was false)

**Comment by ShGKme** [comment]
PR Status update: checking issue with Talk loosing tab navigation with focus trap

**Comment by ShGKme** [comment]
The Talk issue is related but appears in any focus trap container with input with the trailing button visible only on focus.

What happens:
1. <kbd>Tab</kbd> -> Focus on input.
2. Now the "Clear" trailing button is visible.
3. <kbd>Tab</kbd> -> Focus the trailing button.
4. <kbd>Tab</kbd> -> issue
   1. Blurring focus from the trailing button hides it, and removes it from the DOM
   2. `focus-trap` with `MutationObserver` sees it and thinks that the current focused element is removed
   3. Then `focus-trap` moves the focus to the first element of the container

focus

Source:
https://github.com/focus-trap/focus-trap/blob/a49e9b247bb4d15e3a16ddcc9e6b847a62725c77/index.js#L862-L875

This behavior is not configurable.

IMO, Possible solutions could be:
- Fix it on the `focus-trap` side
- Do not remove the trailing button from the DOM, only hide it
- Hide the trailing button with a delay

**Comment by ShGKme** [comment]
Another issue with the `NcAppNavigationItem`. Focus gets stuck on the last item

**Review by susnux — APPROVED** [review]

**Comment by susnux** [comment]
Maybe report this upstream, but

> On Talk: Do not remove the trailing button from the DOM, only hide it

Sounds the best for now.

**Review by GretaD — APPROVED** [review]

**Review by JuliaKirschenheuter — APPROVED** [review]

**Review by AndyScherzinger — APPROVED** [review]

**Pull request merged**